### PR TITLE
KAFKA-6778: AdminClient.describeConfigs() should return error for non-existent topics

### PR DESCRIPTION
**Pull request by omkreddy**

### CODE REVIEW

**Comment by omkreddy** [comment]
@hachikuji  please take a look when you get a chance. Thanks.

File: core/src/main/scala/kafka/server/AdminManager.scala [hunk 1]
```diff
@@ -303,11 +303,15 @@ class AdminManager(val config: KafkaConfig,
...
             val topic = resource.name
-            Topic.validate(topic)
-            // Consider optimizing this by caching the configs or retrieving them from the `Log` when possible
-            val topicProps = adminZkClient.fetchEntityConfig(ConfigType.Topic, topic)
-            val logConfig = LogConfig.fromProps(KafkaServer.copyKafkaConfigToLog(config), topicProps)
-            createResponseConfig(allConfigs(logConfig), createTopicConfigEntry(logConfig, topicProps, includeSynonyms))
+            if (metadataCache.contains(topic)) {
+              Topic.validate(topic)
+              // Consider optimizing this by caching the configs or retrieving them from the `Log` when possible
+              val topicProps = adminZkClient.fetchEntityConfig(ConfigType.Topic, topic)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/AdminManager.scala]: Shouldn't `fetchEntityConfig` indicate that the topic doesn't exist?

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala]: @ijuma  Thanks for the review.  fetchEntityConfig return empty props for non-existing entities. We can add a empty check. But i thought using metadataCache is consistent with other methods in KafkaAPIs. Let me you know your preference.

EDIT: If we add a config and delete config after some time,  we are not deleting config znode.  In this case  fetchEntityConfig will return empty props,  even if topic exists.  So, as of now, we can not use fetchEntityConfig result to check topic existence. We should rely on metadataCache.

**Comment by omkreddy** [comment]
@hachikuji pinging for review

File: core/src/main/scala/kafka/server/AdminManager.scala (hunk 1 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/AdminManager.scala]: I think `fetchEntityConfigs` should return an `Option` and `None` for the case where the node doesn't exist.

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala]: Yes, fetchEntityConfigs can return Optional props. I can update this as part of separate PR.

Comment (ijuma) [reply core/src/main/scala/kafka/server/AdminManager.scala]: I'm suggesting that it's a better approach than the one in the current PR.

Comment (ijuma) [reply core/src/main/scala/kafka/server/AdminManager.scala]: Oh sorry, I had not seen your edit. That is weird that we do not delete the znode. Is there a reason for that? We should file a JIRA. In that case, it does seem like we have to rely on the metadata cache.

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala]: @ijuma  Your suggestion works. We are deleting config znode, when we delete topic. We are not deleting znode, when we don't have any custom configs. 
My only concern is, Do we need to depend on config znode for topic existence check?  while creating topic, Kafka admin script creates empty config znode. Someone may create topic with custom zk script, which may not create config znode while creating topic.

File: core/src/main/scala/kafka/server/AdminManager.scala [hunk 2]
```diff
@@ -303,11 +303,15 @@ class AdminManager(val config: KafkaConfig,
 
           case ResourceType.TOPIC =>
             val topic = resource.name
-            Topic.validate(topic)
-            // Consider optimizing this by caching the configs or retrieving them from the `Log` when possible
-            val topicProps = adminZkClient.fetchEntityConfig(ConfigType.Topic, topic)
-            val logConfig = LogConfig.fromProps(KafkaServer.copyKafkaConfigToLog(config), topicProps)
-            createResponseConfig(allConfigs(logConfig), createTopicConfigEntry(logConfig, topicProps, includeSynonyms))
+            if (metadataCache.contains(topic)) {
+              Topic.validate(topic)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/AdminManager.scala]: Should the topic validation go first? Presumably, if it's in the cache, it must be valid?

Comment (omkreddy) [reply core/src/main/scala/kafka/server/AdminManager.scala]: @hachikuji  Thanks for the review. yes, topic validation should come first. updated the code.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch.

**Pull request merged**